src/libcore/iter.rs

```diff
@@ -369,6 +369,25 @@ pub trait Iterator {
     /// `None` here means that either there is no known upper bound, or the
     /// upper bound is larger than `usize`.
     ///
+    /// # Safety
+    ///
+    /// Implementations of this trait are not guaranteed to yield
+    /// declared number of elements. Iterator may yield less than lower
+    /// bound or more than upper bound of elements.
+    ///
+    /// `size_hint()` can be used to optimize fast path, and can be used
+    /// to panic if estimation is incorrect, but must not be used e. g.
+    /// to omit bound checks in unsafe operations. Incorrect estimation
+    /// of size must not break memory safety.
+    ///
+    /// That said implementation should provide proper estimation,
+    /// because otherwize it may hurt performance or lead to panic.
+    ///
+    /// # Performance
+    ///
+    /// `size_hint()` should be cheap, it should work in `O(1)` time,
+    /// and should be cheaper than call to `next()`.
```

Comment:
@bluss yes, you are right about `log(n)`. I don't know how to say better.

* just "cheap" is vague
* `O(n)` is unacceptable
* `O(1)` is probably two strict
* `O(n / log(n))` is too complicated

I should probably drop Performance section.